numpyro/mcmc.py

```diff
@@ -0,0 +1,124 @@
+        return pack_fn(dist.norm(0., np.sqrt(np.reciprocal(inverse_mass_matrix))).rvs(random_state=rng))
+    elif inverse_mass_matrix.ndim == 2:
+        raise NotImplementedError
+
+
+def hmc_kernel(potential_fn, kinetic_fn):
+    vv_init, vv_update = velocity_verlet(potential_fn, kinetic_fn)
+    unflatten_fn = None
+
+    def init_kernel(init_samples,
+                    num_warmup_steps,
+                    step_size=1.0,
+                    num_steps=None,
+                    adapt_step_size=True,
+                    adapt_mass_matrix=True,
+                    diag_mass=True,
+                    target_accept_prob=0.8,
+                    rng=PRNGKey(0)):
+        assert isinstance(step_size, float)
+        if num_steps is None:
+            trajectory_length = 2 * math.pi
+        else:
+            trajectory_length = num_steps * step_size
+
+        z_flat, pack_fn = ravel_pytree(init_samples)
+        nonlocal unflatten_fn
+        unflatten_fn = pack_fn
+        m_inv_init = np.ones(np.shape(z_flat))
+        rng, rng_momentum = random.split(rng)
+        momentum_generator = partial(_sample_momentum, m_inv_init, pack_fn, rng_momentum)
+        find_initial_ss = partial(find_reasonable_step_size, potential_fn, kinetic_fn, momentum_generator,
+                                  m_inv_init, init_samples)
+
+        wa_init, wa_update = warmup_adapter(num_warmup_steps,
+                                            find_reasonable_step_size=find_initial_ss,
+                                            adapt_step_size=adapt_step_size,
+                                            adapt_mass_matrix=adapt_mass_matrix,
+                                            diag_mass=diag_mass,
+                                            target_accept_prob=target_accept_prob)
+
+        wa_state = wa_init(step_size, mass_matrix_size=np.size(z_flat))
+        r = _sample_momentum(wa_state.inverse_mass_matrix, pack_fn, rng)
+        vv_state = vv_init(init_samples, r)
+
+        def body_fn(t, args):
+            vv_state, wa_state, rng = args
+            rng, rng_momentum, rng_transition = random.split(rng, 3)
+            r = _sample_momentum(wa_state.inverse_mass_matrix, pack_fn, rng_momentum)
+            vv_state = vv_state.update(r=r)
+            num_steps = _get_num_steps(wa_state.step_size, trajectory_length)
+            accept_prob, vv_state_new = _next(num_steps, wa_state.step_size, wa_state.inverse_mass_matrix, vv_state)
+            transition = random.bernoulli(rng_transition, accept_prob)
+            if jax.api._jit_is_disabled:
+                if transition:
+                    vv_state = vv_state_new
+            else:
+                vv_state = lax.cond(transition,
+                                    vv_state_new, lambda state: state,
+                                    vv_state, lambda state: state)
+            z_flat, _ = ravel_pytree(vv_state.z)
+            wa_state = wa_update(t, accept_prob, z_flat, wa_state)
```

Comment:
I haven't looked at this in detail, but should we pass in the proposal state here along with the acceptance probability or the selected state (after the biased coin flip). I have stuck with what we are doing in Pyro, but I am not sure about this.

Reply:
I think we'll put the selected state. It is what Stan does AFAIK. Unfortunately, I couldn't find a reference for it. :(